### PR TITLE
CMake (Windows): Added support for ninja build and some fixes/changes

### PR DESCRIPTION
**Pull request by ChaseOxide**
Most changes have comments stuck with them, but here's the summary of it.  

Changes made:
- Added EXACT flag on find_package CUDA to clarify it is building for exactly that version, the output "minimum required" can be misleading
- Tests if the compiler is compatible with CUDA before it starts building
- BYPRODUCTS/BUILD_BYPRODUCTS are added for Ninja to search for dependency, otherwise it will output the error shown below  
`ninja: error: 'zlib/install/lib/zlibstatic.lib', needed by 'proto_text.exe', missing and no known rule to make it`
- Visual Studio is the only generator with that will output under $(Configuration) directory, thus removed if it is not Visual Studio
- Fixed CONFIGURE_COMMAND causing some CMake versions to break (it resets to default generator, the [CMake](https://cmake.org/files/v3.6/cmake-3.6.3-win64-x64.zip) I tried uses NMake Makefiles as default, which causes the output below  
  ```
  ...
  Submodule path 'third_party/benchmark': checked out '360e66c1c4777c99402cf8cd535aa510fee16573'
  Performing update step for 'protobuf'
  No patch step for 'protobuf'
  Performing configure step for 'protobuf'
  CMake Error at CMakeLists.txt:12 (project):
    Generator

      NMake Makefiles

    does not support platform specification, but platform

      x64

    was specified.

  -- Building for: NMake Makefiles
  -- Configuring incomplete, errors occurred!
  ...
  ```

Additional Info: The current version of CMake has a problem with its ninja generator, as it may produce an error below  
`ninja: fatal: CreateProcess: The filename or extension is too long.`  
The CMake with its ninja generator improved can be found [here](https://gitlab.kitware.com/cmake/cmake/merge_requests/1604)

### CODE REVIEW

**Comment by ChaseOxide** [comment]
I signed it!

File: tensorflow/contrib/cmake/CMakeLists.txt [hunk 1]
```diff
@@ -286,7 +286,21 @@ if (tensorflow_ENABLE_GPU)
     list(APPEND CMAKE_LIBRARY_PATH "${tensorflow_CUDA_LIBRARY_PATH}/stubs")
   endif (NOT WIN32)
 
-  find_package(CUDA ${tensorflow_CUDA_VERSION} REQUIRED)
+  # later command will make use of the value in tensorflow_CUDA_VERSION
+  find_package(CUDA ${tensorflow_CUDA_VERSION} REQUIRED EXACT)
+
+  # Test compatibality of compiler on CUDA
```
Comment (mrry) [inline tensorflow/contrib/cmake/CMakeLists.txt]: Nit (typo): "compatibility"

**Review by mrry — CHANGES_REQUESTED** [review]
Thanks for the contribution! Looks fine to me modulo one typo.

**Comment by mrry** [comment]
@jhseu Before merging this, we should probably run a manual Windows CMake GPU build in addition to the regular presubmits.

**Review by mrry — APPROVED** [review]

**Pull request merged**